### PR TITLE
Refactor gSP.lights and gSP.lookat

### PR DESCRIPTION
**Pull request by gizmo98**
I refactored gSP.lights and gSP.lookat as described here:
https://github.com/gonetz/GLideN64/issues/1447

I also rewrote InverseTransformVectorNormalize 2x and 4x so they can benefit from the new structure. 
Rewritten versions are faster:
2x, opt level -O2
old 0,42 runtime of c function
new 0,36 runtime of c function

4x, opt level  -O2
old 0,30 runtime of c function
new 0,23 runtime of c function

Tested with and without -DVEC4_OPT and -DNEON_OPT.

### CODE REVIEW

**Comment by gonetz** [comment]
~15% speedup. Not bad!

File: src/3DMath.cpp [hunk 1]
```diff
@@ -88,16 +88,16 @@ void Normalize(float v[3])
 #endif // WIN32_ASM
 }
 
-void InverseTransformVectorNormalize2x(float src0[3], float src1[3], float dst0[3], float dst1[3], float mtx[4][4] )
+void InverseTransformVectorNormalize2x(float src[2][3], float dst[2][3], float mtx[4][4] )
 {
-	InverseTransformVectorNormalize(src0, dst0, mtx);
-	InverseTransformVectorNormalize(src1, dst1, mtx);
+	InverseTransformVectorNormalize(&src[0], &dst[0], mtx);
```
Comment (gonetz) [inline src/3DMath.cpp]: Compilation error:

3DMath.cpp:93: error: cannot convert 'float (*)[3]' to 'float*' for argument '1' to 'void InverseTransformVectorNormalize(float*, float*, float (*)[4])'
  InverseTransformVectorNormalize(&src[0], &dst[0], mtx);
                                                       ^

File: src/gSP.h [hunk 2]
```diff
@@ -30,6 +30,9 @@
 #define MODIFY_RGBA	0xFF000000
 #define MODIFY_ALL	0xFFFFFFFF
 
+enum Component { R, G, B };
+enum Axis { X ,Y, Z, W };
+
```
Comment (gonetz) [inline src/gSP.h]: SPLight not used anymore

File: src/3DMath.cpp (hunk 1 above)
Comment (gizmo98) [inline src/3DMath.cpp]: With g++ or vc?

Comment (gizmo98) [inline src/3DMath.cpp]: I had no error with g++ 4.9. I will take a look.

File: src/gSP.h [hunk 3]
```diff
@@ -76,8 +79,24 @@ struct gSPInfo
...
-	SPLight lights[12];
-	SPLight lookat[2];
+	struct
+	{
+		f32 rgb[12][3];
+		f32 xyz[12][3];
+		f32 i_xyz[12][3];
+		f32 pos_xyzw[12][4];
+		f32 ca[12], la[12], qa[12];
+	}lights;
```
Comment (gonetz) [inline src/gSP.h]: space between } and identifier

File: src/gSP.h [hunk 4]
```diff
@@ -76,8 +79,24 @@ struct gSPInfo
...
+
+	struct
+	{
+		f32 rgb[2][3];
+		f32 xyz[2][3];
+		f32 i_xyz[2][3];
+		f32 pos_xyzw[2][4];
+		f32 ca[2], la[2], qa[2];
+	}lookat;
```
Comment (gonetz) [inline src/gSP.h]: space between } and identifier

File: src/gSP.cpp [hunk 5]
```diff
@@ -344,7 +344,7 @@ void gSPProcessVertex4(u32 v)
 				f32 x, y;
 				if (gSP.lookatEnable) {
 					x = DotProduct(gSP.lookat.i_xyz[0], fLightDir);
-					y = DotProduct(gSP.lookat.i_xyz[0], fLightDir);
+					y = DotProduct(gSP.lookat.i_xyz[1], fLightDir);
```
Comment (gonetz) [inline src/gSP.cpp]: to first commit

File: src/3DMath.cpp (hunk 1 above)
Comment (gonetz) [inline src/3DMath.cpp]: g++ (Ubuntu 4.8.4-2ubuntu1~14.04.3) 4.8.4

Comment (gonetz) [inline src/3DMath.cpp]: Visual says:
3dmath.cpp(93): error C2664: 'void InverseTransformVectorNormalize(float [],float [],float [][4])': cannot convert argument 1 from 'float (*)[3]' to 'float []'
3>        gliden64\src\3dmath.cpp(93): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast

**Comment by gizmo98** [comment]
Ok. I fixed the error.

**Comment by gonetz** [comment]
Thanks! Merged: b8a18f57e464

**Pull request closed without merging**